bokeh/server/protocol/receiver.py

```diff
@@ -0,0 +1,118 @@
+''' Assemble websocket wire message fragments into complete Bokeh Server
+message objects that can be processed.
+
+'''
+from __future__ import absolute_import
+
+from six import b
+from tornado.concurrent import return_future
+
+from ..exceptions import MessageError, ProtocolError, ValidationError
+
+HMAC           = 0
+HEADER         = 1
+METADATA       = 2
+CONTENT        = 3
+BUFFER_HEADER  = 4
+BUFFER_PAYLOAD = 5
+
+class Receiver(object):
+    '''
+
+    On MessageError or ValidationError, the receiver will reset its state
+    and attempt to consume a new message.
+
+    [
+        # these are required
+        b'foobarbaz',       # HMAC signature
+        b'{header}',        # serialized header dict
+        b'{metadata}',      # serialized metadata dict
+        b'{content},        # serialized content dict
+
+        # these are optional, and come in pairs
+        b'{buf_header}',    # serialized buffer header dict
+        b'array'            # raw buffer payload data
+        ...
+    ]
+
+    '''
+
+    def __init__(self, protocol):
+        self._protocol = protocol
+        self._expecting = HMAC
+        self._message = None
+        self._buf_header = None
+        self._failures = 0
+
+    @return_future
```

Comment:
It seems `return_future` is Tornado's "old way" of doing event-based programming. Can't you use a `gen.coroutine` instead?